Adding support for Gerrit group operations

sstafford opened the pull request:
I have a need to perform operations on Gerrit groups, but there is currently no support in the API.  I've implemented support for many of the operations defined in the GroupAPI.  One thing I wasn't able to get working is support for parsing the "visibile_to_all" option.  You'll see this as a failing unit test, so hopefully someone knows how to correctly parse the project options.  (see ./src/test/resources/com/urswolfer/gerrit/client/rest/http/groups/groups.json for an example of the "visible_to_all" option)

uwolfer commented:
Thank you very much for your PR.

I will do a very brief review for the moment for basic things which should help you to get things moved forward:
* Please do not create "manual" JSON (like in #addMembers). Use Java data structures and convert them to JSON with Gson.
* All API (and API documentation) needs to go into `com.google.gerrit.extensions.api` package. Please see other API there. Once it is in a good shape, we can contribute it to official Gerrit (I can support you / contribute it to Gerrit for you).
* Use only one empty separator line (e.g. between methods).
* Use `Url#encode` (instead of your own #getUrlEncoded).
* `visible_to_all`: I assume the JSON used in your test is copied from a real Gerrit instance? I don't see the mistake right now - will need to look into it more deeply.
* Use your name as author in new files. And probably attach your cooperate email address to your GitHub account ;)

Side note: Please update your pull request by amending your initial commit and force-push it to your remote. That way the PR gets updated.

sstafford commented:
I believe I've fixed the items other than the "visible_to_all" issue.

coffeecup-winner commented on src/test/resources/com/urswolfer/gerrit/client/rest/http/groups/groups.json:
```diff
@@ -0,0 +1,63 @@
+    "id": "834ec36dd5e0ed21a2ff5d7e2255da082d63bbd7",
+    "url": "#/admin/groups/uuid-834ec36dd5e0ed21a2ff5d7e2255da082d63bbd7",
+    "options": {
+      "visible_to_all": true,
```
This line breaks the build - you should remove the trailing comma.

uwolfer replied:
@coffeecup-winner: Great catch - thanks!

uwolfer commented on src/main/java/com/urswolfer/gerrit/client/rest/GerritApiImpl.java:
```diff
@@ -43,6 +46,16 @@
     private final GerritRestClient gerritRestClient;
 
+    /**
+     * Construct the REST API for managing Group information.
```
I think you can omit this comment since this is quite boilerplate code (it's the same for all APIs below).

uwolfer commented on src/main/java/com/urswolfer/gerrit/client/rest/http/groups/GroupApiRestClient.java:
```diff
@@ -0,0 +1,253 @@
+public class GroupApiRestClient implements GroupApi {
+
+    /** Base REST URL for managing Group data */
+    public static final String BASE_URL = "/groups";
```
this can be private

uwolfer commented on src/main/java/com/urswolfer/gerrit/client/rest/http/groups/GroupApiRestClient.java:
```diff
@@ -0,0 +1,253 @@
+    /** Base REST URL for managing Group data */
+    public static final String BASE_URL = "/groups";
+
+
```
This empty line and the one on line 43 can be removed in order to follow other code style.

sstafford replied:
BASE_URL is referenced from another class (GroupsRestClient) because the URL being constructed in the create() method needs to it.  I'll make the variable private, but will need to add a method in GroupApiRestClient for accessing the information.

uwolfer commented on src/main/java/com/urswolfer/gerrit/client/rest/http/groups/GroupApiRestClient.java:
```diff
@@ -0,0 +1,253 @@
+        this.groupId = id;
+    }
+
+    public static String getRequestPath(String id) {
```
private

uwolfer commented on src/main/java/com/urswolfer/gerrit/client/rest/http/groups/GroupApiRestClient.java:
```diff
@@ -0,0 +1,253 @@
+    }
+
+    /** @return group info with no {@code ListGroupsOption}s set. */
+    public GroupInfo get() throws RestApiException {
```
Please add @Override annotations to all methods which are implementing / override methods.

uwolfer commented on src/main/java/com/urswolfer/gerrit/client/rest/http/groups/GroupApiRestClient.java:
```diff
@@ -0,0 +1,253 @@
+        return getRequestPath(groupId);
+    }
+
+    /** @return group info with no {@code ListGroupsOption}s set. */
```
IMHO this documentation (and some docs below) can be removed since it is already documented in the interface.

uwolfer commented on src/main/java/com/urswolfer/gerrit/client/rest/http/groups/GroupApiRestClient.java:
```diff
@@ -0,0 +1,253 @@
+     */
+    public void owner(String owner) throws RestApiException {
+        String restPath = getRequestPath() + "/owner";
+        JsonElement result = gerritRestClient.putRequest(restPath, owner);
```
Variable result can be removed.

uwolfer commented on src/main/java/com/urswolfer/gerrit/client/rest/http/groups/GroupApiRestClient.java:
```diff
@@ -0,0 +1,253 @@
+     * @return group members.
+     * @throws RestApiException
+     */
+    public List<AccountInfo> members(boolean recursive) throws RestApiException {
```
Argument recursive is ignored below. Would be nice if you could cover this in a test.

sstafford replied:
I don't have any experience with mock frameworks, so I just tried to fix the code so it didn't ignore the method argument.  I wasn't sure about your preference for creating a new string or appending to a StringBuffer, so I didn't use the append approach.

uwolfer commented on src/main/java/com/urswolfer/gerrit/client/rest/http/groups/GroupApiRestClient.java:
```diff
@@ -0,0 +1,253 @@
+        memberMap.put("members", Arrays.asList(members));
+        String json = gerritRestClient.getGson().toJson(memberMap);
+
+        JsonElement result = gerritRestClient.postRequest(restPath, json);
```
Variable result is also not necessary here (and in some other places in this file) - please remove it.

uwolfer commented on src/main/java/com/urswolfer/gerrit/client/rest/http/groups/GroupsParser.java:
```diff
@@ -0,0 +1,63 @@
+ */
+public class GroupsParser {
+    private static final Type GROUP_MAP_TYPE = new TypeToken<SortedMap<String, GroupInfo>>() {}.getType();
+    private static final Type GROUP_LIST_TYPE = new TypeToken<List<GroupInfo>>() {}.getType();
```
Unused at the moment (commented out below).

uwolfer commented on src/main/java/com/urswolfer/gerrit/client/rest/http/groups/GroupsParser.java:
```diff
@@ -0,0 +1,63 @@
+
+        //list = gson.fromJson(result, GROUP_LIST_TYPE);
+        SortedMap<String, GroupInfo> map = gson.fromJson(result, GROUP_MAP_TYPE);
+        list = new ArrayList(map.values());
```
Variable can be declared here. And please initialize it with type (ArrayList<GroupInfo>).

uwolfer commented on src/main/java/com/urswolfer/gerrit/client/rest/http/groups/GroupsParser.java:
```diff
@@ -0,0 +1,63 @@
+    public List<AccountInfo> parseGroupMembers(JsonElement result) throws RestApiException {
+        List<AccountInfo> list = null;
+        if (!result.isJsonArray()) {
+            // Plase the single element in a list
```
Typo ("Plase")?

uwolfer commented on src/main/java/com/urswolfer/gerrit/client/rest/http/groups/GroupsRestClient.java:
```diff
@@ -0,0 +1,117 @@
+                List<GroupInfo> list = GroupsRestClient.this.list(this);
+                if (list != null) {
+                    Iterator listIter = list.iterator();
+                    while (listIter.hasNext()) {
```
I prefer a foreach here (for (GroupInfo group : list) {...), but it's up to you :)

uwolfer commented on src/main/java/com/urswolfer/gerrit/client/rest/http/groups/GroupsRestClient.java:
```diff
@@ -0,0 +1,117 @@
+    private List<GroupInfo> list(ListRequest listParameter) throws RestApiException {
+        List<GroupInfo> list = null;
+
+        // TODO Implement the other query parameters supported by Groups.ListRequest
```
If you do not implemented it yet, please throw an NotImplementedException when they are passed. Otherwise people are wondering why it does nothing when they pass them.

sstafford replied:
I don't have any experience with mock frameworks, so I just tried to fix the code so it didn't ignore the method argument.  I wasn't sure about your preference for creating a new string or appending to a StringBuffer, so I didn't use the append approach.

uwolfer commented on src/main/java/com/urswolfer/gerrit/client/rest/http/groups/GroupsRestClient.java:
```diff
@@ -0,0 +1,117 @@
+        }
+        JsonElement result = gerritRestClient.getRequest(url);
+        if (result == null) {
+            list = new ArrayList<GroupInfo>();
```
Collections.emptyList()

uwolfer commented on src/test/java/com/urswolfer/gerrit/client/rest/http/groups/GroupApiRestClientTest.java:
```diff
@@ -0,0 +1,23 @@
+
+public class GroupApiRestClientTest {
+
+
```
Please remove this 2nd empty line.

uwolfer commented on src/test/java/com/urswolfer/gerrit/client/rest/http/groups/GroupsParserTest.java:
```diff
@@ -0,0 +1,58 @@
+import java.util.List;
+
+/**
+ * @author Thomas Forrer
```
Your name here :)

uwolfer commented on src/test/java/com/urswolfer/gerrit/client/rest/http/groups/GroupApiRestClientTest.java:
```diff
@@ -0,0 +1,23 @@
+package com.urswolfer.gerrit.client.rest.http.groups;
+
+import com.google.common.truth.Truth;
```
A lot of these imports are unused.

uwolfer commented on src/test/java/com/urswolfer/gerrit/client/rest/http/groups/GroupApiRestClientTest.java:
```diff
@@ -0,0 +1,23 @@
+public class GroupApiRestClientTest {
+
+
+    private GroupsRestClient getGroupsRestClient(GerritRestClient gerritRestClient) {
```
Where do you want to use this method? It seems to be unused.

uwolfer commented on src/test/java/com/urswolfer/gerrit/client/rest/http/groups/GroupsParserTest.java:
```diff
@@ -0,0 +1,58 @@
+import com.urswolfer.gerrit.client.rest.http.common.GerritAssert;
+import org.testng.annotations.Test;
+
+import java.util.Collections;
```
Unused.

uwolfer commented on src/test/java/com/urswolfer/gerrit/client/rest/http/groups/GroupsRestClientTest.java:
```diff
@@ -0,0 +1,57 @@
+import org.testng.annotations.Test;
+
+/**
+ * @author Thomas Forrer
```
Your name here.

uwolfer commented on src/test/java/com/urswolfer/gerrit/client/rest/http/groups/GroupsRestClientTest.java:
```diff
@@ -0,0 +1,57 @@
+        return gerritRestClient;
+    }
+
+    private GerritRestClient gerritRestClientExpectPut(String expectedUrl) throws Exception {
```
Unused.

uwolfer commented on src/test/java/com/urswolfer/gerrit/client/rest/http/groups/GroupsRestClientTest.java:
```diff
@@ -0,0 +1,57 @@
+        return gerritRestClient;
+    }
+
+    private GerritRestClient gerritRestClientExpectDelete(String expectedUrl) throws Exception {
```
Unused.

uwolfer commented:
Thanks again for your work. Once this PR is merged, I can push a new release (incl. Maven Repo) if you like.

Btw, I would prefer if you could squash commits (you can do that when everything is ready if you like). You can just force-push the squashed commit to your GitHub branch which will replace all existing commits.

sstafford commented:
I think I've addressed all of the comments...

1. Commits have been squashed into a single commit.  Sorry about that, I thought I had been amending all along.  Hopefully it came out ok...

2. GroupApiRestClient.members() - I wasn't sure how to test this.  Added my comments as a reply to yours.

3. GroupsRestClient.list(ListRequest) - Again, this would also need tests but my knowledge of mock isn't sufficient to do it at the moment.  My comments are provided at that line.

4.  I just use the Atom editor, so I didn't catch a lot of the issues that you pointed out.  It might be good to add some maven static code analysis plugins (checkstyle, findbugs, pmd, or something) to help out here.  Let me know if you're interested in that and I'll create a separate pull request.

uwolfer commented:
> Commits have been squashed into a single commit. Sorry about that, I thought I had been amending all along. Hopefully it came out ok...

Looks good to me.

> GroupApiRestClient.members() - I wasn't sure how to test this. Added my comments as a reply to yours.

I have changed it a bit. Will push the merge commit including some minor changes soon.

> GroupsRestClient.list(ListRequest) - Again, this would also need tests but my knowledge of mock isn't sufficient to do it at the moment. My comments are provided at that line.

I have added some more tests. Will push them soon so you can see a way how to implement it ;)

> I just use the Atom editor, so I didn't catch a lot of the issues that you pointed out. It might be good to add some maven static code analysis plugins (checkstyle, findbugs, pmd, or something) to help out here. Let me know if you're interested in that and I'll create a separate pull request.

Yes, that would be nice. Probably it could also be checked with Travis CI which is used for this project. PRs are very welcome.

Btw, you have not added any copyright headers to new files. Do you agree to add the same copyright headers as used in other files? I will do that in my merge commit if you agree.

sstafford commented:
Yes, the same copyright headers as the other files is fine.

uwolfer commented:
@sstafford: If your are interested how I have written more tests for your change, please check out my last commit.

If you need an updated Maven release, please post a comment.
